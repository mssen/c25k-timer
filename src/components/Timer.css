.current-time {
  font-size: 100px;
}

.total-time {
  font-size: 40px;
  color: gray;
}