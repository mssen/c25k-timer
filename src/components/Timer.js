import React from 'react';

// Yes, this is basically the same thing as the stateful component example from: https://facebook.github.io/react/
class Timer extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      seconds: 0,
      intervalIndex: 0
    }
  }

  tick() {
    const nextSeconds = this.state.seconds + 1;

    // End of interval
    if (nextSeconds > this.props.intervals[this.state.intervalIndex]) {

      // End of segment
      if (this.state.intervalIndex === this.props.intervals.length - 1) {
        clearInterval(this.interval);

      // On to the next segment
      } else {
        this.setState((prevState) => ({
          seconds: 1,
          intervalIndex: prevState.intervalIndex + 1
        }));
      }

    // Business as usual
    } else {
      this.setState({
        seconds: nextSeconds
      });
    }
  }

  componentDidMount() {
    this.interval = setInterval(() => this.tick(), 1000);
  }

  componentWillUnmount() {
    clearInterval(this.interval);
  }

  format(seconds) {
    const remainderSeconds = seconds % 60;
    return Math.floor(seconds / 60) + ':' + (remainderSeconds < 10 ? '0' : '') + remainderSeconds;
  }

  render() {
    return (
      <div>
        { `${this.format(this.state.seconds)} / ${this.format(this.props.intervals[this.state.intervalIndex])}` }
      </div>
    );
  }
}

export default Timer;
