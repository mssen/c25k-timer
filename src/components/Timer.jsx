import React from 'react';

// Yes, this is basically the same thing as the stateful component example from: https://facebook.github.io/react/
class Timer extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      seconds: 0,
      intervalIndex: 0
    }
  }

  tick() {
    const nextSeconds = this.state.seconds + 1;

    // End of interval
    if (nextSeconds > this.props.intervals[this.state.intervalIndex]) {

      // End of segment
      if (this.state.intervalIndex === this.props.intervals.length - 1) {
        this.stop();

      // On to the next segment
      } else {
        const newIntervalIndex = this.state.intervalIndex + 1;
        this.setState(({
          seconds: 1,
          intervalIndex: newIntervalIndex
        }));
        this.props.updateInterval(newIntervalIndex);
      }

    // Business as usual
    } else {
      this.setState({
        seconds: nextSeconds
      });
    }
  }

  start() {
    this.interval = setInterval(() => this.tick(), 1000);
  }

  stop() {
    clearInterval(this.interval);
    this.interval = false;
  }

  componentDidMount() {
    this.props.play && this.start();
  }

  componentWillUnmount() {
    this.stop();
  }

  componentWillReceiveProps(nextProps) {
    if (nextProps.segment !== this.props.segment) {
      this.setState({
        seconds: 0,
        intervalIndex: 0
      });
    }
    if (!this.interval && nextProps.play) {
      this.start();
    }
    if (this.interval && !nextProps.play) {
      this.stop();
    }
  }

  format(seconds) {
    const remainderSeconds = seconds % 60;
    return Math.floor(seconds / 60) + ':' + (remainderSeconds < 10 ? '0' : '') + remainderSeconds;
  }

  render() {
    return (
      <div>
        { `${this.format(this.state.seconds)} / ${this.format(this.props.intervals[this.state.intervalIndex])}` }
      </div>
    );
  }
}

export default Timer;
