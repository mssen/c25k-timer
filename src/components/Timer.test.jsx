import React from 'react';
import { shallow, mount } from 'enzyme';
import Timer from './Timer';

jest.useFakeTimers();

it('renders without crashing', () => {
  shallow(
    <Timer
      intervals={[10]}
      updateInterval={() => ''}
      segment={0}
      play={false}
    />
  );
});

it('is paused when created', () => {
  const wrapper = mount(
    <Timer
      intervals={[10]}
      updateInterval={() => ''}
      segment={0}
      play={false}
    />
  );
  expect(wrapper).toMatchSnapshot();
  jest.runTimersToTime(1000);
  expect(wrapper).toMatchSnapshot();
});

it('can start', () => {
  const wrapper = mount(
    <Timer intervals={[10]} updateInterval={() => ''} segment={0} play={true} />
  );
  expect(wrapper).toMatchSnapshot();
  jest.runTimersToTime(1000);
  expect(wrapper).toMatchSnapshot();
});

it('can pause', () => {
  const wrapper = mount(
    <Timer intervals={[10]} updateInterval={() => ''} segment={0} play={true} />
  );
  expect(wrapper).toMatchSnapshot();
  jest.runTimersToTime(1000);
  expect(wrapper).toMatchSnapshot();
  wrapper.setProps({ play: false });
  jest.runTimersToTime(1000);
  expect(wrapper).toMatchSnapshot();
});

it('switches over to the next segment', () => {
  const wrapper = mount(
    <Timer intervals={[10]} play={true} segment={1} updateInterval={() => ''} />
  );
  jest.runTimersToTime(1000);
  expect(wrapper).toMatchSnapshot();
  wrapper.setProps({ segment: 2, intervals: [5] });
  expect(wrapper).toMatchSnapshot();
});

describe('next interval', () => {
  it('goes on to the next interval', () => {
    const updateIntervalStub = jest.fn();
    const wrapper = mount(
      <Timer
        intervals={[5, 5]}
        play={true}
        updateInterval={updateIntervalStub}
        segment={0}
      />
    );
    expect(wrapper).toMatchSnapshot();
    jest.runTimersToTime(2000);
    expect(wrapper).toMatchSnapshot();
    jest.runTimersToTime(4000);
    expect(wrapper).toMatchSnapshot();
    expect(updateIntervalStub).toHaveBeenCalledWith(1);
  });

  it('stops when it was the last interval', () => {
    const updateIntervalStub = jest.fn();
    const wrapper = mount(
      <Timer
        intervals={[5, 5]}
        play={true}
        updateInterval={updateIntervalStub}
        segment={0}
      />
    );
    jest.runTimersToTime(12000);
    expect(wrapper).toMatchSnapshot();
    expect(updateIntervalStub).toHaveBeenCalledWith(1);
  });
});
