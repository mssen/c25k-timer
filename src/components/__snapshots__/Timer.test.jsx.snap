// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`can pause 1`] = `
<Timer
  intervals={
    Array [
      10,
    ]
  }
  play={true}
  segment={0}
  updateInterval={[Function]}
>
  <div>
    <span
      className="current-time"
    >
      0:00
    </span>
    <span
      className="total-time"
    >
      /
      0:10
    </span>
  </div>
</Timer>
`;

exports[`can pause 2`] = `
<Timer
  intervals={
    Array [
      10,
    ]
  }
  play={true}
  segment={0}
  updateInterval={[Function]}
>
  <div>
    <span
      className="current-time"
    >
      0:01
    </span>
    <span
      className="total-time"
    >
      /
      0:10
    </span>
  </div>
</Timer>
`;

exports[`can pause 3`] = `
<Timer
  intervals={
    Array [
      10,
    ]
  }
  play={false}
  segment={0}
  updateInterval={[Function]}
>
  <div>
    <span
      className="current-time"
    >
      0:01
    </span>
    <span
      className="total-time"
    >
      /
      0:10
    </span>
  </div>
</Timer>
`;

exports[`can start 1`] = `
<Timer
  intervals={
    Array [
      10,
    ]
  }
  play={true}
  segment={0}
  updateInterval={[Function]}
>
  <div>
    <span
      className="current-time"
    >
      0:00
    </span>
    <span
      className="total-time"
    >
      /
      0:10
    </span>
  </div>
</Timer>
`;

exports[`can start 2`] = `
<Timer
  intervals={
    Array [
      10,
    ]
  }
  play={true}
  segment={0}
  updateInterval={[Function]}
>
  <div>
    <span
      className="current-time"
    >
      0:01
    </span>
    <span
      className="total-time"
    >
      /
      0:10
    </span>
  </div>
</Timer>
`;

exports[`is paused when created 1`] = `
<Timer
  intervals={
    Array [
      10,
    ]
  }
  play={false}
  segment={0}
  updateInterval={[Function]}
>
  <div>
    <span
      className="current-time"
    >
      0:00
    </span>
    <span
      className="total-time"
    >
      /
      0:10
    </span>
  </div>
</Timer>
`;

exports[`is paused when created 2`] = `
<Timer
  intervals={
    Array [
      10,
    ]
  }
  play={false}
  segment={0}
  updateInterval={[Function]}
>
  <div>
    <span
      className="current-time"
    >
      0:00
    </span>
    <span
      className="total-time"
    >
      /
      0:10
    </span>
  </div>
</Timer>
`;

exports[`next interval goes on to the next interval 1`] = `
<Timer
  intervals={
    Array [
      5,
      5,
    ]
  }
  play={true}
  segment={0}
  updateInterval={[Function]}
>
  <div>
    <span
      className="current-time"
    >
      0:00
    </span>
    <span
      className="total-time"
    >
      /
      0:05
    </span>
  </div>
</Timer>
`;

exports[`next interval goes on to the next interval 2`] = `
<Timer
  intervals={
    Array [
      5,
      5,
    ]
  }
  play={true}
  segment={0}
  updateInterval={[Function]}
>
  <div>
    <span
      className="current-time"
    >
      0:02
    </span>
    <span
      className="total-time"
    >
      /
      0:05
    </span>
  </div>
</Timer>
`;

exports[`next interval goes on to the next interval 3`] = `
<Timer
  intervals={
    Array [
      5,
      5,
    ]
  }
  play={true}
  segment={0}
  updateInterval={[Function]}
>
  <div>
    <span
      className="current-time"
    >
      0:01
    </span>
    <span
      className="total-time"
    >
      /
      0:05
    </span>
  </div>
</Timer>
`;

exports[`next interval stops when it was the last interval 1`] = `
<Timer
  intervals={
    Array [
      5,
      5,
    ]
  }
  play={true}
  segment={0}
  updateInterval={[Function]}
>
  <div>
    <span
      className="current-time"
    >
      0:05
    </span>
    <span
      className="total-time"
    >
      /
      0:05
    </span>
  </div>
</Timer>
`;

exports[`switches over to the next segment 1`] = `
<Timer
  intervals={
    Array [
      10,
    ]
  }
  play={true}
  segment={1}
  updateInterval={[Function]}
>
  <div>
    <span
      className="current-time"
    >
      0:01
    </span>
    <span
      className="total-time"
    >
      /
      0:10
    </span>
  </div>
</Timer>
`;

exports[`switches over to the next segment 2`] = `
<Timer
  intervals={
    Array [
      5,
    ]
  }
  play={true}
  segment={2}
  updateInterval={[Function]}
>
  <div>
    <span
      className="current-time"
    >
      0:00
    </span>
    <span
      className="total-time"
    >
      /
      0:05
    </span>
  </div>
</Timer>
`;
